tools/cli.js

```diff
@@ -24,14 +24,64 @@ const log = (msg) => {
   if (userInput) rl.write(userInput);
 };
 
+function completeByFields(input, object) {
+  return Object.keys(object)
+    .filter(c => !c.startsWith('_') && (!input || c.startsWith(input)));
+}
+
+function tryComplete(input, completer) {
+  const completions = completer._complete([input], 0)[0];
+  if (completions.length === 1) return completions[0];
+  return input;
+}
+
+function completer(line) {
+  const inputs = _split(line, ' ', 0, true);
+  const [completions, help] = iterativeCompletion(inputs, 0, commandProcessor);
+  if (help) log('\n' + help);
+  // to allow partial completion, as method above gives
+  // completions for the latest command part
+  const lastPart = inputs.length === 0 ? line : inputs[inputs.length - 1];
+  return [completions, lastPart];
+}
+
+// inputs - array of user inputs
+// depth - level of nested completion (index in inputs array)
+// completer - object that has '_complete(inputs, depth)' function or ._help()
+//             or neither (no completions or help available)
+function iterativeCompletion(inputs, depth, completer) {
+  function helper(depth, oldDepth, completer, completions) {
+    let help = '';
+
+    if (completions.length !== 1) return [completions, help];
+    const nextCompleter = completer[completions[0]];
+    if (!nextCompleter) return [completions, help];
+
+    if (nextCompleter._complete && depth < inputs.length) {
+      const [newCompletions, newDepth] = nextCompleter._complete(inputs, depth);
+      return helper(newDepth, depth, nextCompleter, newCompletions);
+    }
+    if (inputs[oldDepth] === completions[0]) {
+      completions.shift();
+      if (nextCompleter._help) help = nextCompleter._help();
+    }
+    return [completions, help];
+  }
+  const [newCompletions, newDepth] = completer._complete(inputs, depth);
```

Comment:
The doc comment above this function states:

> // completer - object that has '_complete(inputs, depth)' function or ._help()
> //             or neither (no completions or help available)

But if the `completer` object has no `_complete` method, this line will fail with a `TypeError`.